[FLINK-24703][connectors][formats] Add CSV format support for filesystem based on StreamFormat and BulkWriter interfaces.

afedulov opened the pull request:
This PR adds a new implementation of CSV format for filesystem based on StreamFormat, BulkWriter interfaces.

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Adds the ability to read and write CSV data using more up-to-date APIs.


## Brief change log

  - Adds new `CsvFormat` based on  `StreamFormat` from `file-connector-files` for reading CSV data from filesystem.
  - Introduces generic `Converter` interface for enabling conversion between Java and SQL data types within the new format.
  - Adds new `CsvBulkWriter` for writing resulting into CSV files.

## Verifying this change

This change added tests and can be verified as follows:

  - Integration tests for reading CSV files as Tables and writing resulting Tables back as CSV files.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no )
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( yes )
  - The serializer: ( yes )
  - The runtime per-record code paths (performance sensitive): ( yes )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? ( JavaDocs )

slinkydeveloper commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/DelimitedFormat.java:
```diff
@@ -0,0 +1,131 @@
+ * with their internal buffering of stream input and charset decoder state.
+ */
+@PublicEvolving
+public class DelimitedFormat<T> extends SimpleStreamFormat<T> {
```
I'm wondering, what is the difference between this format and `DeserializationSchemaAdapter`? Could the latter be dropped in favour of this new class?

afedulov replied:
This is an obsolete approach, please ignore. This PR is still WIP and not ready for review.

afedulov commented:
@flinkbot run azure

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFormatFactory.java:
```diff
@@ -19,103 +19,48 @@
- * and {@link DeserializationSchema}.
- */
+/** CSV format factory for file system. */
+//TODO: Rename back to CsvFileSystemFormatFactory? @Arvid
```
Why do you think so?
Can we find a name that doesn't start with Csv so it pops up later in auto-complete? `TableCsv...` or `InternalCsv...`?

afedulov replied:
I had this thought because `BulkReader` is a `flink-connector-file` interface, and `CsvBulkWriter` writer into `FSDataOutputStream`. Both the encoder and the decoder exposed by the factory are for files.
I like the `TableCsvFormatFactory`, renamed.

afedulov replied:
I guess we could pass the `OutputStream` instead of the  `FSDataOutputStream`, but that would mean adding a check via `isinstanceof` in the `finish()` method for calling `sync()` for files.

afedulov replied:
I like the `TableCsvFormatFactory` - will rename it.

afedulov replied:
I forgot that the `BulkWriter` factory is passed an `FSDataOutputStream`, so technically both ins and outs of this factory are only for supporting files.

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvToRowDataConverters.java:
```diff
@@ -65,7 +66,12 @@ public CsvToRowDataConverters(boolean ignoreParseErrors) {
      */
     @FunctionalInterface
-    public interface CsvToRowDataConverter extends Serializable {
+    public interface CsvToRowDataConverter extends Converter<JsonNode, Object, Void> {
```
Make sure to remove as much `public` as possible from non-public stuff.

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvToRowDataConverters.java:
```diff
@@ -305,7 +311,7 @@ private CsvToRowDataConverter createArrayConverter(ArrayType arrayType) {
 
     private static void validateArity(int expected, int actual, boolean ignoreParseErrors) {
-        if (expected != actual && !ignoreParseErrors) {
+        if (expected > actual && !ignoreParseErrors) {
```
Can you double-check if this is still needed?

afedulov replied:
The `CsvFormat.Reader` read() method needs to read elements according to the full schema from the file but emmit elements with a potentially reduced number of rows according to the projection pushdown.

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,89 @@
+    private final Converter<T, JsonNode, RowDataToCsvFormatConverterContext> converter;
+    private final ObjectWriter csvWriter;
+    /** Reusable within the converter. * */
+    private final transient ObjectNode container;
```
Drop transient.

afedulov replied:
This is in accordance with the design choice explicitly made in a similar situation in the recent addition:

https://github.com/apache/flink/commit/b1e436a109c6473b794f02ec0a853d1ae6df6c83#diff-58a105b8d92eadfb0af487ba6b2bce5a00955fceb2077e6b7f2d1fa01592a50bR84

One could argue that as it is a throw-away object that does not itself contain any configuration there is no strict necessity to ship it to the executors. Let me know what you think and if there are general guidelines for similar cases.

AHeise replied:
You'd add `transient` only for `Serializable` classes. `BulkWriter` is not while `SerializationSchema` is. (BulkWriter is the better design)

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,89 @@
+     *     elements of type JsonNode.
+     * @param stream The output stream.
+     */
+    public CsvBulkWriter(
```
Make package-private and add factory methods.

AHeise commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFormat.java:
```diff
@@ -0,0 +1,144 @@
+     * @param pojoType The type class of the POJO.
+     * @param <T> The type of the returned elements.
+     */
+    public static <T> CsvFormat<T> from(Class<T> pojoType) {
```
Make sure this method is covered by a test and feels natural to DataStream/java dev.

AHeise commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatITCase.java:
```diff
@@ -0,0 +1,204 @@
+import static org.apache.flink.table.utils.DateTimeUtils.unixTimestampToLocalDateTime;
+
+/** Tests for the CSV file format. */
+public class CsvFormatITCase extends JsonPlanTestBase {
```
Add these tests to the respective production commits. If it's only reader, then into the first commit. If it's writer or both, then second commit.

AHeise commented on flink-formats/flink-csv/src/test/resources/META-INF/services/org.apache.flink.table.factories.Factory:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+org.apache.flink.formats.csv.CsvFormatFactory
+org.apache.flink.formats.csv.CsvSerDeSchemaFactory
```
This should also be in earlier commits.

AHeise commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatITCase.java:
```diff
@@ -0,0 +1,204 @@
+public class CsvFormatITCase extends JsonPlanTestBase {
+
+    @Test
+    public void testProjectPushDown() throws Exception {
```
Here there are no DataStream tests whatsoever. Please add them to cover your new functionality (don't bother covering old stuff).

afedulov replied:
Do you mean testing the new `CsvBulkWriter` and `CsvReaderFormat` not anything related to the ser/de schema-based (old) stuff? Another question - you point at the projection pushdown test, not sure if you meant it explicitly (new functionality?) or if it is a coincidence. Do we even have facilities for communicating pushdown from the streaming API?

AHeise replied:
> Do you mean testing the new CsvBulkWriter + CsvReaderFormat and not anything related to the ser/de schema-based (old) stuff?

Yes

> Another question - you point at the projection pushdown test, not sure if you meant it explicitly (new functionality?) or if it is a coincidence. Do we even have facilities for communicating pushdown from the streaming API?

I can only write comments on changed lines. So it's a coincidence.

AHeise reviewed (commented):
Looking good already. I have added a few minor comments and nits but also 2 larger fields:
- Please make sure that all classes are annotated (not just new) or hidden (non-public). Get rid of as many public classes as possible (so the user has less to choose from).
- There is not a single test looking at DataStream. Please add respective tests so that we know how the new format classes feel. Make sure to offer factories for BulkWriter.

slinkydeveloper commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFormatFactory.java:
```diff
@@ -125,128 +70,88 @@ public String factoryIdentifier() {
+            final DataType projectedDataDype = Projection.of(projections).project(physicalDataType);
+            final RowType rowTypeProjected = (RowType) projectedDataDype.getLogicalType();
+
+            final RowType rowType = (RowType) physicalDataType.getLogicalType();
```
Name it `physicalRowType` so it makes the distinction with the above variable more clear

afedulov replied:
How about `getLogicalType()`? Would `logicalRowType` be more appropriate?

slinkydeveloper replied:
no, because `RowType` is a `LogicalType`. `physicalRowType` makes clear that this one is the physical type, and not the projected one

afedulov replied:
Thanks,  makes sense.

slinkydeveloper commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvSerDeSchemaFactory.java:
```diff
@@ -0,0 +1,192 @@
+
+        return new DecodingFormat<DeserializationSchema<RowData>>() {
+            @Override
+            public DeserializationSchema<RowData> createRuntimeDecoder(
```
I wonder if this one can be a projectable decoding format as well, given it roughly uses the same code for deserializing.

afedulov replied:
I guess it could. 
@AHeise, I believe we did not previously discuss this aspect. Do you think it is necessary to expand the scope of projection pushdown support also to the Ser/De schema-based implementation for this PR?

AHeise replied:
This would be orthogonal and can be done later. I'm also not sure if it's worth the hassle now: performance gain on a per-record projection is comparably small and we implicitly have it on all formats with reader schema.

However, I'd like you to verify if the current version actually works with projections, so can you please add an appropriate test?

slinkydeveloper replied:
@AHeise But then why `CsvFormatFactory` returns a `ProjectableDecodingFormat`? It doesn't make sense, becuse either none of them are projectable or both are.

AHeise replied:
Is this fully supported in Table-land? If so, then I agree we should change it. It's a good performance-improvement.

slinkydeveloper replied:
> Is this fully supported in Table-land? If so, then I agree we should change it. It's a good performance-improvement.

Yes, please check the documentation of `DecodingFormat` which goes through all the implementation details both for formats and connectors

afedulov replied:
@slinkydeveloper 
I am not sure about this actually. The focus of this PR is 
**Add CSV format support for filesystem based on StreamFormat and BulkWriter interfaces.**
I would rather not touch the existing Ser/De schema-based functionality in this scope. Nothing will break if we won't support pushdown in the non-filesystem connectors for now and to me, it sounds like an improvement that should be done in a separate PR.

afedulov replied:
Notice that `CsvSerDeSchemaFactory` is just a rename of the previously existing `CsvFormatFactory`, we did not touch its functionality in the PR.

slinkydeveloper replied:
Ok but then open a followup issue for that, because this inconsistency between the `CsvFormatFactory` and `CsvSerDeSchemaFactory` is confusing and doesn't really make sense to exist, as either both support projections, or none of them does.

afedulov replied:
@slinkydeveloper as discussed, https://issues.apache.org/jira/browse/FLINK-25931.

slinkydeveloper commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvFormatITCase.java:
```diff
@@ -0,0 +1,204 @@
+                sinkPath);
+    }
+
+    private static String formatSqlTimestamp(long timestamp) {
+        return TimeFormats.SQL_TIMESTAMP_FORMAT.format(unixTimestampToLocalDateTime(timestamp));
+    }
```
Use `formatUnixTimestamp(long unixtime, TimeZone tz)` instead, as it also makes sure you use a fixed timezone and you don't get weird time zone issues when running on other machines :)

afedulov replied:
Not sure what you mean. I can see the `formatUnixTimestamp`  only in JavaScript libraries in Flink's code base. Or is your point that the timezone should necessarily be also taken into account?

slinkydeveloper replied:
The method I'm talking about should be in `DateTimeUtils` in table-common

afedulov replied:
Found this addition after rebase, thanks.

afedulov replied:
I will skip this one - the test data from the SQL suite is created based on local datetime, so the tests break if another zone is used anyhow: https://github.com/apache/flink/blob/a7192af8707f3f0d0f30fc71f3477edd92135cac/flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/utils/TestData.scala#L267
I do not see it as an issue, because 
1) No absolut values are used - only relative since 1970 for both the test data and the expected results.
2) The local time zone will this way always be used and I do not see how it can break on another machine because it using a different timezone.

slinkydeveloper commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/JsonPlanTestBase.java:
```diff
@@ -92,17 +92,27 @@ protected void createTestValuesSourceTable(
+            @Nullable String partitionFields,
+            Map<String, String> properties) {
+        checkArgument(fieldNameAndTypes.length > 0);
+        String partitionedBy =
+                StringUtils.isNullOrWhitespaceOnly(partitionFields)
+                        ? ""
+                        : "\n partitioned by (" + partitionFields + ") \n";
```
please remove this function usage and rewrite it using `Stream`s instead

slinkydeveloper reviewed (commented):
I have a couple of minor comments, otherwise looks good from table point of view

afedulov commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFormat.java:
```diff
@@ -0,0 +1,144 @@
+    // ------------------------------------------------------------------------
+
+    /** The actual reader for the {@code CsvFormat}. */
+    public static final class Reader<R, T> implements StreamFormat.Reader<T> {
```
Good point. Copy-paste error from the `TextLineFormat` format "template".

afedulov commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFormat.java:
```diff
@@ -0,0 +1,144 @@
+
+/**
+ * A {@code StreamFormat} for reading CSV files.
+ *
```
Do you have a good inspirational example for this kind of instruction in our codebase?

afedulov replied:
@fapaul maybe you have a good example?

afedulov commented:
@AHeise @slinkydeveloper thank you for the feedback.
I addressed your comments where there were no clarifications needed. I am still working on a couple more tests for the DataStream API, but apart from that this PR is ready for another pass.

AHeise reviewed (commented):
Almost there. Will do another round later when the writer test is there.

afedulov commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,289 @@
+import static org.junit.Assert.assertTrue;
+
+/** MiniCluster-based integration tests CSV data format. */
+public class DataStreamCsvITCase extends TestLogger {
```
Our current `CsvBulkWriter` is currently pinned to be used in the Table/SQL context because it relies on the `RowDataToCsvFormatConverterContext`. It seems that we might need a different approach given that using it in the DataStream API is also a requirement.
https://github.com/apache/flink/pull/17598/files#diff-a32dfa08e3b14134b196779c88529037e79fd77d350de57d7738dcd4d295f9eaR81

afedulov replied:
@AHeise I decoupled the writer from `RowData` in a separate hotfix [`8074cb4` (#17598)](https://github.com/apache/flink/pull/17598/commits/8074cb4ace3b77d71d67926d289d7322ff3eb4c2)
It makes instantiation in the factory a bit awkward, but it works. What do you think?
If we keep it, how about putting a Serializable requirement also explicitly on the context type of the Converter interface?
https://github.com/apache/flink/pull/17598/files#diff-eeda89bce5e097c1019fe734e379cc5fb7a2c2f5e528d5b26bc715118eece38eR30

afedulov replied:
@fapaul I've added the explicit `Serializable` requirement to the `Converter` interface, because to me it looks like it is required. What do you think?

AHeise commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -82,41 +102,86 @@ public void testBoundedTextFileSourceWithJackson() throws Exception {
+        return contents;
+    }
+
+    static final class TestCsvBulkWriterFactory implements BulkWriter.Factory<CityPojo> {
```
This feels a bit awkward, how about

```
class CsvBulkWriter {
  static <T> BulkWriter.Factory<T> forPojo(Class<T> pojoClass) {
  
            final ObjectMapper objectMapper = new ObjectMapper();
            PojoConverter.ConverterContext converterContext =
                    new PojoConverter.ConverterContext(objectMapper);
            final PojoConverter converter = createConverter(); // see next comment
            
            final CsvMapper csvMapper = new CsvMapper();
            final CsvSchema schema = csvMapper.schemaFor(pojoClass);
    new CsvBulkWriter.Factory(schema, csvMapper, converter); // basically same ctor as CsvFormat with converter going into the opposite direction
  }
}

afedulov replied:
Added native POJOs support in CsvBulkWriter.

AHeise commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -82,41 +102,86 @@ public void testBoundedTextFileSourceWithJackson() throws Exception {
+        }
+    }
+
+    interface PojoConverter extends Converter<CityPojo, JsonNode, PojoConverter.ConverterContext> {
+
+        class ConverterContext implements Serializable {
+            ObjectMapper mapper;
+
+            public ConverterContext(ObjectMapper csvMapper) {
+                this.mapper = csvMapper;
+            }
+        }
+    }
+
+    public static PojoConverter createConverter() {
+        return (pojo, context) -> context.mapper.valueToTree(pojo);
```
Why do we need a special PojoConverter? On reading side, we just do `(value, context) -> value` for POJO, right? Shouldn't the converter already receive the mapped value?

afedulov replied:
Added native POJOs support in CsvBulkWriter.

AHeise reviewed (commented):
Yes, I thought something was missing. Going into the right direction. Let's try to make Writer and Reader symmetric.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/TableCsvFormatFactory.java:
```diff
@@ -0,0 +1,195 @@
+        private final ReadableConfig formatOptions;
+
+        public CsvBulkDecodingFormat(ReadableConfig formatOptions) {
+            this.formatOptions = formatOptions;
```
nit: checkNotNull

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,351 @@
+
+    @ClassRule public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();
+
+    @ClassRule
```
Can you change the newly added tests to use junit 5 and assertJ?

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,351 @@
+    // ------------------------------------------------------------------------
+    //  test cases
+    // ------------------------------------------------------------------------
+    @Test
```
Do you also want to add an ITCase for `CsvReaderFormat.fromSchema`?

afedulov replied:
It is actually also tested implicitly via `fromPojo` https://github.com/afedulov/flink/blob/02355cdca67d73500f5251834bc07df18ec1f2a8/flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvReaderFormat.java#L109
Apart from the native schema construction via Jackson there seems not much be changed in the additional test. Or do you see it differently?

afedulov replied:
I've added an explicit test.

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,351 @@
+                            .setNumberSlotsPerTaskManager(PARALLELISM)
+                            .setRpcServiceSharing(RpcServiceSharing.DEDICATED)
+                            .withHaLeadershipControl()
+                            .build());
```
I am a bit confused by this class. Is this mini-cluster even used? After looking more thoroughly this class seems more like a normal unit test rather than an IT case

fapaul replied:
If you want to write an IT case you can use the new File. I hope with 1.15 we can deprecate `StreamingFileSink`.

afedulov replied:
Good point, I did not realize that `StreamExecutionEnvironment.getExecutionEnvironment();` does not need it.

fapaul replied:
Ah, now I see what happened here :) It is actually correct to keep the mini cluster but I would suggest making the second test either in a separate test class because it is not an IT case or also converting it to an IT case.

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,351 @@
+        assertThat(result, is(CSV_LINES));
+    }
+
+    static <T> BulkWriter.Factory<T> factoryForPojo(Class<T> pojoClass) {
```
private?

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,351 @@
+        return (out) -> new CsvBulkWriter<>(csvMapper, schema, converter, null, out);
+    }
+
+    static Map<File, String> getFileContentByPath(File directory) throws IOException {
```
private?

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/TableCsvFormatITCase.java:
```diff
@@ -0,0 +1,204 @@
+/** Tests for the CSV file format. */
+public class TableCsvFormatITCase extends JsonPlanTestBase {
+
+    @Test
```
Same as for the other test class can you use junit 5 and assertJ, please.

afedulov replied:
This test is largely based on utilities from the `JsonPlanTestBase` and `AbstractTestBase` base classes. I would prefer to leave the migration of those utility classes for table API to someone who works on Table/SQL. What do you think?

fapaul replied:
We will probably take over all testing utils for formats. The right way would be to convert the `JsonPlanTestBase` into a junit extension and get rid of the inheritance.

I can understand if you do not want to do it as part of this PR but eventually with the next format we have to do it :)

afedulov replied:
Yeah,  `AbstractTestBase` which is JUnit4-based is used in 400 classes, I propose to address it in a separate effort :)

fapaul replied:
I mean you do not have to touch `AbstractTestBase` it does not offer much besides a MiniCluster which you can also get with the existing extension. The only effort would be to rework `JsonPlanTestBase`.

afedulov replied:
`JsonPlanTestBase` is also used in 28 tests :sweat_smile: . I created a separate `TableFormatTestBase` based on JUnit5 and migrated my test to it.

slinkydeveloper replied:
Please let's keep `JsonPlanTestBase` and don't add a new test base. The test base you added is based on legacy as well (`TestValuesTableFactory`), and it's doing assertions based on strings while we want to use the new `TableAssertions`. So your new base will immediately become just another legacy to clean up later :smile:. 

We're planning to do proper rework of our bases and push them in the new `flink-table-test-utils`, and these needs to be independent of table planner, so connector and formats stop depending on table planner test classpath. If you wanna create a new test base, feel free to start from this point: push it to `flink-table-test-utils` and use the new `TableAssertions`. But I'm pretty sure it's quite some work, so no need to do it in this PR.

afedulov replied:
Reverted back to the `JsonPlanTestBase`.

fapaul commented on flink-formats/flink-format-common/src/main/java/org/apache/flink/formats/common/Converter.java:
```diff
@@ -0,0 +1,39 @@
+ * @param <To> The output type.
+ * @param <C> The context for passing optional conversion instructions.
+ */
+public interface Converter<From, To, C> extends Serializable {
```
I guess this interface is supposed to be `@PublicEvolving` or `@Experimental` since it might be used in the future by a lot of formats.

afedulov replied:
Good point, is there a preference which one to use in this particular case?

fapaul replied:
I think you have to mark it is as `@PublicEvolving` because it is used by your new CSV format that also is `@PublicEvolving`

fapaul reviewed (commented):
Looks mostly good I left some comments regarding the test setup and some minor cleanup.

afedulov commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,107 @@
+            FSDataOutputStream stream) {
+        this.converter = converter;
+        this.converterContext = converterContext;
+        this.stream = stream;
```
As far as the documentation states, it is supposed to be handled by the new sink machinery
https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-core/src/main/java/org/apache/flink/api/common/serialization/BulkWriter.java#L69

afedulov commented:
@fapaul thanks a lot for your feedback. I've addressed your comments, apart from the IT/Unit-test related question. Actually, I am bit confused with that regard - my implementation is based largely on these `FileSource` tests:
https://github.com/apache/flink/blob/575aa2dbfb25af65436d60dd8ed2c0b1d0aa2ef8/flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/src/FileSourceTextLinesITCase.java#L126
However, the tests still work as expected after removing the `MiniCluster` from the test completely.

afedulov commented:
@fapaul I addressed all of the comments. The only remaining open points from my perspective:

- clarify if imposing a strict `Serializable` requirement on the `Converter` interface is desirable. I believe Arvid mentioned that shifting this concern to factories was described as "better design" (we do not have factories for the existing converters though).
- clarify if there should also be an explicit `CsvWriterFactory` implementation of `BulkWriter.Factory<T>` for consistency (similar to `ParquetWriterFactory<T>` and `AvroWriterFactory<T>`)
- open a follow-up ticket for projection pushdown support for the previously existing ser/de-schema-based implementation

fapaul commented on flink-formats/flink-csv/pom.xml:
```diff
@@ -77,6 +77,14 @@ under the License.
 
 		<!-- test dependencies -->
 
+		<dependency>
```
Nit: Why was needed?

afedulov replied:
Good point, it is a leftover from an earlier testing approach, not needed anymore.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,114 @@
+            Converter<T, R, C> converter,
+            @Nullable C converterContext,
+            FSDataOutputStream stream) {
+        checkNotNull(mapper);
+        checkNotNull(schema);
+        checkNotNull(converter);
+        checkNotNull(stream);
```
Nit: `checkNotNull` returns the value if not null so you could do:

```java
this.converter = checkNotNull(converter)
```

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,114 @@
+
+    private final FSDataOutputStream stream;
+    private final Converter<T, R, C> converter;
+    private final C converterContext;
```
Nit: annotate with nullable

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvCommons.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    private static void validateCharacterVal(
+            ReadableConfig tableOptions, ConfigOption<String> option, boolean unescape) {
+        if (tableOptions.getOptional(option).isPresent()) {
```
Nit: I prefer explicit condition handling to improve readability.

```java
if (!tableOptions.getOptional(option).isPresent()) {
    return;
}
...
```

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvReaderFormat.java:
```diff
@@ -0,0 +1,171 @@
+ * low-level {@code forSchema} static factory method based on the {@code Jackson} library utilities:
+ *
+ * <pre>{@code
+ * CsvReaderFormat<T> forSchema(CsvMapper mapper,
```
Nit: The example does not look very helpful. Maybe you can show a code snipped in conjunction with using the format with the FileSource and it would also be great two a second example in the same code block showing it for a POJO class.

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/** MiniCluster-based integration tests CSV data format. */
+public class DataStreamCsvITCase extends TestLogger {
```
You are using the `TestLogger` for junit4. Please replace it with the `TestLoggerExtension` [1]

[1] https://github.com/apache/flink/blob/3fe1de954bf3ee05247cdff0c43d71dc9535803d/flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaCommitterTest.java#L43

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+
+    @TempDir File outDir;
+
+    public static final MiniClusterExtension MINI_CLUSTER_RESOURCE =
```
Does this need to be public?

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+                            .build());
+
+    @RegisterExtension
+    public static AllCallbackWrapper<MiniClusterExtension> allCallbackWrapper =
```
Public required?

afedulov replied:
Got some wrong "inspirations"
https://github.com/apache/flink/blob/49a05f4fb94370dd04480c16b04732c8a854cbbb/flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskCancelAsyncProducerConsumerITCase.java#L80

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+                "Beijing,39.905,116.3914,China,CN,Beijing,primary,19433000"
+            };
+
+    final CityPojo[] pojos =
```
Nit: static

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+    //  test data
+    // ------------------------------------------------------------------------
+
+    private static final String[] CSV_LINES =
```
Can you move all static variables to the beginning of the class? The variable declaration should be always at the top of the file that other developers do not have to search them first. You can still generate the test data in a method and put it here.

afedulov replied:
Moved static. I don't quite understand what you have in mind with regarding the method.

fapaul replied:
No worries I meant something like

```java
private static final String[] CSV_LINES = getCsvLines();
```
to prevent having a lot of lines for the initialization just for next time :)

fapaul reviewed (commented):
Looks almost ready. I left some last cleanup tasks and currently, the CI fails due to some checkstyle violations.

Can you address all the remaining points, squash all commits, and rebase with the latest master?

afedulov commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,363 @@
+
+        String[] result = getResultsFromSinkFiles(outDir);
+
+        assertThat(result).containsExactlyInAnyOrder(CSV_LINES);
```
Not sure what you mean, the call is checking that the lines are there in any order (containsExactlyIn*Any*Order)

afedulov commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/TableCsvFormatITCase.java:
```diff
@@ -0,0 +1,204 @@
+public class TableCsvFormatITCase extends JsonPlanTestBase {
+
+    @Test
+    public void testProjectPushDown() throws Exception {
```
I've added support for the new implementation of the formats:
https://github.com/afedulov/flink/blob/744b699a82b4640652050eb75eecaf78b2f10c2c/flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java#L104
The discussion was about also backporting/adding this feature to the already-existing SerDe-schema-based implementation within the scope of this PR.

afedulov commented:
Thanks @fapaul, I addressed your comments, squashed the hotfix commits and rebased on master.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvBulkWriter.java:
```diff
@@ -0,0 +1,112 @@
+
+    private final FSDataOutputStream stream;
+    @Nullable private final Converter<T, R, C> converter;
+    private final C converterContext;
```
I think the wrong variable is now annotated nullable.

afedulov replied:
Oops.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvCommons.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    public static Set<ConfigOption<?>> optionalOptions() {
```
Can you align the options with this PR https://github.com/apache/flink/pull/18290 ?

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,390 @@
+        }
+
+        final File parent = file.getParentFile();
+        assertTrue(parent.mkdirs() || parent.exists());
+
+        assertTrue(stagingFile.renameTo(file));
```
Please use assertj

fapaul commented on flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java:
```diff
@@ -0,0 +1,390 @@
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Please only use assertJ

fapaul reviewed (approved):
LGTM please check the last nitty comments.

afedulov commented:
@fapaul 
Done. 
I've added two hotfix commits that I would like you to give a look at before I squash them.
1. Adds support for the missing `ignore-parse-errors` option. Some tests were failing because of that. It is a bit of a [special option](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/formats/csv/#csv-ignore-parse-errors) when compared to the parsing options of Jackson, that's why I decided to make it a setter not to "pollute" all of the factory methods with it.
2. Adds the `forwardOptions` from Francesco's PR. I believe that is all that is needed, but please confirm.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvFileFormatFactory.java:
```diff
@@ -109,17 +110,21 @@ public CsvBulkDecodingFormat(ReadableConfig formatOptions) {
                             converter,
-                            context.createTypeInformation(projectedDataType)));
+                            context.createTypeInformation(projectedDataType));
+            csvReaderFormat.setIgnoreParseErrors(ignoreParseErrors);
```
Nit: WDYT about passing the complete configuration to the `Reader` class? It feels more extensible in case we also need to pass other configurations to the `Reader`.

The setter, in this case, is only invoked one time and it is similar to a builder method.

afedulov replied:
It is a bit unclear what do you propose? Do you mean adding something like a `ReaderConfig` with a builder and passing it into the `CsvReaderFormat` instead of the individual parameters? 
As mentioned in my previous comment, I decided to make it a setter because otherwise, we would need to expand all of the static factory methods (`fromShcma`, `fromPojo`) and "pollute" them with this special concern. In that case I would add utility "telescope" methods with this parameter set to "ignore-parse-errors=false" by default.

fapaul replied:
Good point. I think the best way here is to use the `wither` pattern [1]. You can add a method like

`public CsvReaderFormat<T> withIgnoreParseErrors()`

that creates a new `CsvReaderFormat` with the flag that it ignores the errors. By solely using `fromSchema` or `fromPojo` nothing changes. Users can now do `CsvReaderFormat.fromPojo(Pojo.class).withIgnoreParseErrors()`

[1] https://projectlombok.org/features/With

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvReaderFormat.java:
```diff
@@ -155,16 +169,34 @@
+                    } else {
+                        return null;
+                    }
+                } catch (Throwable t) {
```
Is it necessary to catch all kinds of throwables here or is `IOException` enough? It can be dangerous to catch all kinds of exceptions i.e. if you handle the `InterruptedException` you also want to set the interrupted flag of the Thread.

afedulov replied:
I decided not to reinvent the wheel and handled it exactly as this option is currently handled in the existing implementation to avoid surprises for the users that already rely on this functionality:
https://github.com/apache/flink/blob/16452fbb57c11418796912251152947601572106/flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDataDeserializationSchema.java#L157

fapaul replied:
Ok let's keep it.

afedulov replied:
But good point.

fapaul commented on flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvReaderFormat.java:
```diff
@@ -155,16 +169,34 @@
         @Override
         public T read() throws IOException {
-            return iterator.hasNext() ? converter.convert(iterator.next(), null) : null;
+            while (true) {
```
Why do you now use a while loop?

afedulov replied:
An arbitrary number of lines can be malformed. Let's say this is a single CSV file
```
a,b,c    // header
1,2,3    // expected
1,          // malformed
1,2,3,4 // malformed
5,6,7   // expected
```
Should produce [1,2,3] and [5,6,7] and ignore the rest. The number of malformed elements is unknown.

fapaul replied:
I see then I'd propose to rewrite the loop to 

```java
while (iterator.hasNext()) {
    try {
        R nextElement = iterator.next();
        return converter.convert(nextElement, 
    } catch (....) {
    }
}
return null;
```

fapaul reviewed (commented):
Thanks for aligning the options. I think you need to resolve some conflicts now and some of the ignore-parse-errors logic looks odd PTAL.

Once you have resolved the comments please squash the fix commits and rebase with the latest master.

afedulov commented:
@fapaul Done. I've also added an explicit test for the `withIgnoreParseErrors` option.

fapaul reviewed (approved):
LGTM
